File: harness/oracle/cases/help-doc-chunking.ts
/**
 * P4.D77 oracle case — help-document chunking (v4 `lib/help/help-doc-chunking.ts`
 * `buildHelpDocChunks` + `helpChunkEmbeddingText`, new at `24633026`).
 *
 * Tier 1: pure functions over a fixed corpus. Imports v4's REAL module (which
 * in turn pulls in the REAL Scriptorium chunker, `lib/mount-index/chunker.ts`)
 * and emits one NDJSON row per input:
 *   { kind: 'chunks', name, input, output: [{ chunkIndex, heading, content }] }
 *   { kind: 'embedText', name, docTitle, heading, content, output }
 *
 * The corpus deliberately probes what a port gets wrong:
 *   - the SIZE TARGETS. 400/700/100 is not the chunker's 800/1200/200 default,
 *     and a port that forgot to pass the options produces a single chunk where
 *     v4 produces five. The multi-section documents below are sized to sit
 *     between the two regimes, so a defaulted port diverges immediately.
 *   - the OVERLAP prefix carried into each chunk after the first, and the
 *     word-boundary trim inside it.
 *   - heading tracking across sections, including a document whose first
 *     paragraphs precede any heading (leading chunks carry `null`).
 *   - the oversized single paragraph (no blank lines at all) that the chunker
 *     must break on sentence/hard boundaries rather than paragraph ones.
 *   - the whitespace-only and empty arms, which yield NOTHING (not one empty
 *     chunk).
 *   - CRLF and a `#`-through-`######` heading ladder, since the heading regex
 *     is line-anchored.
 *   - ⚠ `helpChunkEmbeddingText`'s guard is `heading ? … : docTitle` — JS
 *     TRUTHINESS, so an EMPTY-STRING heading takes the title-only branch. A
 *     port matching on `Some(_)` passes every other case and fails only here.
 *   - the U+203A separator itself (a single ›, not '>' and not ' > ').
 *
 * Run (Node 24, from the v4 checkout):
 *   N=~/.nvm/versions/node/v24.13.1/bin ; V5=<this worktree>
 *   cd ~/source/quilltap-server
 *   $N/node --import tsx $V5/harness/oracle/cases/help-doc-chunking.ts \
 *     > /tmp/oracle-help-doc-chunking.ndjson
 */

/** A filler paragraph of a known length, repeated to reach a target size. */
const FILLER = 'Words about the subject at hand, repeated for length. '

/** v4's own test helper, verbatim — sections of `## Section N` + filler. */
function longDocument(sections: number): string {
  const filler = FILLER.repeat(60)
  return Array.from({ length: sections }, (_, i) => `## Section ${i}\n\n${filler}`).join('\n\n')
}

const CHUNK_CORPUS: Array<[string, string]> = [
  // Degenerate: nothing at all comes back (not one empty chunk).
  ['empty', ''],
  ['whitespace-only', '   \n\n  '],
  ['newlines-only', '\n\n\n'],

  // A short document stays whole, and carries its H1 as the heading.
  ['short-with-h1', '# Title\n\nA brief paragraph.'],
  ['short-no-heading', 'Just a sentence, with no heading above it.'],

  // ⚠ THE SIZE-TARGET PROBE. ~2,600 chars: over the help 700-token (2,800-char)
  // ceiling is what matters — this one sits just under it so it stays whole,
  // and its bigger sibling below crosses it while staying UNDER the
  // Scriptorium's 1,200-token (4,800-char) default. A port that forgot to pass
  // HELP_CHUNK_OPTIONS keeps the sibling whole and diverges there.
  ['just-under-help-ceiling', `## Heading\n\n${FILLER.repeat(45)}`],
  ['over-help-under-scriptorium', `## Heading\n\n${FILLER.repeat(75)}`],

  // Multi-section documents: sequential indices, per-section headings, overlap.
  ['two-sections', longDocument(2)],
  ['six-sections', longDocument(6)],
  ['eight-sections', longDocument(8)],

  // Prose BEFORE the first heading — the leading chunks must carry null.
  [
    'preamble-then-headings',
    `${FILLER.repeat(40)}\n\n## First\n\n${FILLER.repeat(40)}\n\n## Second\n\n${FILLER.repeat(40)}`,
  ],

  // One oversized paragraph, no blank lines: the hard-split path.
  ['single-huge-paragraph', FILLER.repeat(120).trim()],

  // The heading ladder (# … ######) and a non-heading '#x'.
  [
    'heading-ladder',
    ['# One', 'body', '## Two', 'body', '### Three', 'body', '###### Six', 'body', '#NotAHeading', 'body'].join(
      '\n\n',
    ),
  ],

  // CRLF line endings across a section boundary.
  ['crlf-sections', `## Alpha\r\n\r\n${FILLER.repeat(40)}\r\n\r\n## Beta\r\n\r\n${FILLER.repeat(40)}`],

  // Non-BMP content, so any UTF-16-vs-scalar slicing difference shows up.
  ['emoji-sections', `## Emoji \u{1F600}\n\n${FILLER.repeat(40)}\u{1F600}\n\n## Plain\n\n${FILLER.repeat(40)}`],

  // A real help document's shape: frontmatter already stripped by the sync.
  [
    'realistic-help-page',
    [
      '# Chat Settings',
      'Everything about how a conversation is run.',
      '## Image Description Settings',
      FILLER.repeat(30),
      '## Uncensored fallback profile',
      FILLER.repeat(30),
      '## Timeouts',
      FILLER.repeat(30),
    ].join('\n\n'),
  ],
]

const EMBED_CORPUS: Array<[string, string, string | null | undefined, string]> = [
  ['title-and-heading', 'Chat Settings', 'Image Description Settings', 'Body text.'],
  ['null-heading', 'Chat Settings', null, 'Body text.'],
  ['undefined-heading', 'Chat Settings', undefined, 'Body text.'],
  // ⚠ THE TRUTHINESS EDGE: '' is falsy, so title-only.
  ['empty-heading', 'Chat Settings', '', 'Body text.'],
  ['whitespace-heading', 'Chat Settings', ' ', 'Body text.'],
  ['empty-title', '', 'Some Heading', 'Body text.'],
  ['empty-content', 'Chat Settings', 'Some Heading', ''],
  ['multiline-content', 'Chat Settings', 'Some Heading', 'Line one.\n\nLine two.'],
  ['separator-lookalikes', 'A > B', '>', 'Body.'],
  ['non-bmp', 'Guide \u{1F600}', 'Section \u{1F600}', 'Body \u{1F600}.'],
]

async function main(): Promise<void> {
  const { buildHelpDocChunks, helpChunkEmbeddingText } = await import('@/lib/help/help-doc-chunking')

  for (const [name, input] of CHUNK_CORPUS) {
    const output = buildHelpDocChunks(input)
    process.stdout.write(JSON.stringify({ kind: 'chunks', name, input, output }) + '\n')
  }

  for (const [name, docTitle, heading, content] of EMBED_CORPUS) {
    process.stdout.write(
      JSON.stringify({
        kind: 'embedText',
        name,
        docTitle,
        // `undefined` would vanish from the JSON; tag it so the Rust side can
        // tell the two falsy-but-different inputs apart.
        heading: heading === undefined ? null : heading,
        headingUndefined: heading === undefined,
        content,
        output: helpChunkEmbeddingText(docTitle, heading, content),
      }) + '\n',
    )
  }

  process.exit(0)
}

main().catch((err) => {
  process.stderr.write(`help-doc-chunking oracle failed: ${err?.stack ?? err}\n`)
  process.exit(1)
})

File: crates/quilltap-core/src/services/help_doc_chunking.rs
//! Help-document chunking — port of v4 `lib/help/help-doc-chunking.ts` (v4
//! `24633026`).
//!
//! Slices a help document into section-sized pieces for embedding. v4's *why*,
//! carried forward: help docs are long and topically broad — `help/chat-settings.md`
//! alone covers a dozen unrelated subsystems — so a single whole-document
//! embedding is a smear that matches any specific question only weakly. Chunking
//! gives each section its own vector, which is what lets "how do I describe an
//! image for a model that can't see?" land on the paragraph that answers it.
//!
//! **The chunker itself is the Scriptorium's** ([`super::mount_index::chunker`],
//! v4 `lib/mount-index/chunker.ts`) — same Markdown-aware paragraph
//! accumulation, same heading tracking, same overlap. v4 reuses it; so does
//! this. Only the size targets differ.

use super::mount_index::chunker::{chunk_document, ChunkOptions};
use crate::db::help_doc_chunks::HelpDocChunkDraft;

/// v4 `HELP_CHUNK_OPTIONS` — deliberately smaller than the Scriptorium's
/// 800–1200 defaults.
///
/// v4's *why*: help sections are short — a settings subsection runs a few
/// hundred words — and the whole point here is precision, so a chunk that
/// swallows four unrelated settings sections defeats the exercise. The overlap
/// keeps a paragraph that straddles a boundary reachable from both sides.
pub const HELP_CHUNK_OPTIONS: ChunkOptions = ChunkOptions {
    target_min_tokens: Some(400),
    target_max_tokens: Some(700),
    overlap_tokens: Some(100),
};

/// v4 `buildHelpDocChunks` — split a help document into chunk drafts, in
/// document order; empty for empty input.
///
/// The title is NOT prepended here — it is added at embedding time
/// ([`help_chunk_embedding_text`]), so the stored chunk text stays a faithful
/// excerpt of the document, which is what gets shown back to a reader.
///
/// @param content the document body, frontmatter already stripped
pub fn build_help_doc_chunks(content: &str) -> Vec<HelpDocChunkDraft> {
    chunk_document(content, HELP_CHUNK_OPTIONS)
        .into_iter()
        .map(|chunk| HelpDocChunkDraft {
            chunk_index: chunk.chunk_index,
            heading: chunk.heading_context,
            content: chunk.content,
        })
        .collect()
}

/// v4 `helpChunkEmbeddingText` — the text actually handed to the embedding
/// provider for a chunk.
///
/// Title and heading are prefixed so a chunk carries the context a reader would
/// have from the page around it — "Uncensored fallback profile" means little on
/// its own, but a great deal under "Chat Settings › Image Description Settings".
///
/// ⚠ v4's guard is `heading ? … : docTitle` — a **JS truthiness** test, so an
/// EMPTY-STRING heading takes the title-only branch just as `null`/`undefined`
/// do. `Some(String::new())` must therefore behave like `None` here; a plain
/// `match heading { Some(h) => … }` would silently emit `"Title › \n\nbody"`.
/// The chunker can produce an empty heading (`## ` followed by whitespace never
/// matches, but a heading whose text trims away does not arise) — the guard is
/// carried because v4's is, not because a caller is known to hit it.
pub fn help_chunk_embedding_text(doc_title: &str, heading: Option<&str>, content: &str) -> String {
    let path = match heading {
        Some(h) if !h.is_empty() => format!("{doc_title} \u{203a} {h}"),
        _ => doc_title.to_string(),
    };
    format!("{path}\n\n{content}")
}

#[cfg(test)]
mod tests {
    use super::*;

    /// v4's test helper `longDocument(sections)`, verbatim in shape.
    fn long_document(sections: usize) -> String {
        let filler = "Words about the subject at hand, repeated for length. ".repeat(60);
        (0..sections)
            .map(|i| format!("## Section {i}\n\n{filler}"))
            .collect::<Vec<_>>()
            .join("\n\n")
    }

    #[test]
    fn empty_input_yields_nothing() {
        assert!(build_help_doc_chunks("").is_empty());
        assert!(build_help_doc_chunks("   \n\n  ").is_empty());
    }

    #[test]
    fn short_document_stays_one_chunk() {
        let chunks = build_help_doc_chunks("# Title\n\nA brief paragraph.");
        assert_eq!(chunks.len(), 1);
        assert_eq!(chunks[0].chunk_index, 0);
        assert!(chunks[0].content.contains("A brief paragraph."));
    }

    #[test]
    fn long_document_splits_with_sequential_indices_and_headings() {
        let chunks = build_help_doc_chunks(&long_document(6));
        assert!(chunks.len() > 1);
        for (i, chunk) in chunks.iter().enumerate() {
            assert_eq!(chunk.chunk_index, i as i64);
            assert!(
                chunk
                    .heading
                    .as_deref()
                    .is_some_and(|h| h.starts_with("Section")),
                "chunk {i} heading {:?}",
                chunk.heading
            );
        }
    }

    #[test]
    fn chunks_stay_small_enough_to_be_worth_the_exercise() {
        // v4's assertion: 700-token target at ~4 chars/token plus the overlap
        // prefix — the point is that one chunk never swallows the document.
        for chunk in build_help_doc_chunks(&long_document(8)) {
            assert!(chunk.content.len() < 4200, "{}", chunk.content.len());
        }
    }

    #[test]
    fn embedding_text_prefixes_title_and_heading() {
        assert_eq!(
            help_chunk_embedding_text(
                "Chat Settings",
                Some("Image Description Settings"),
                "Body text."
            ),
            "Chat Settings \u{203a} Image Description Settings\n\nBody text."
        );
        assert_eq!(
            help_chunk_embedding_text("Chat Settings", None, "Body text."),
            "Chat Settings\n\nBody text."
        );
        // The JS-truthiness edge: an empty heading is falsy, so title-only.
        assert_eq!(
            help_chunk_embedding_text("Chat Settings", Some(""), "Body text."),
            "Chat Settings\n\nBody text."
        );
    }
}
